src/hardware/adlib.cpp

```diff
@@ -164,7 +165,48 @@ struct Handler : public Adlib::Handler {
 
 }
 
+namespace NukedOPL {
 
+struct Handler : public Adlib::Handler {
+	opl3_chip chip = {};
+	Bit8u newm = 0;
+
+	void WriteReg(Bit32u reg, Bit8u val) override
+	{
+		OPL3_WriteRegBuffered(&chip, (Bit16u)reg, val);
+		if (reg == 0x105)
+			newm = reg & 0x01;
+	}
+
+	Bit32u WriteAddr(Bit32u port, Bit8u val) override
+	{
+		Bit16u addr;
+		addr = val;
+		if ((port & 2) && (addr == 0x05 || newm)) {
+			addr |= 0x100;
+		}
+		return addr;
+	}
+
+	void Generate(MixerChannel *chan, Bitu samples) override
+	{
+		Bit16s buf[1024 * 2];
+		while (samples > 0) {
+			Bit32u todo = samples > 1024 ? 1024 : samples;
+			samples -= todo;
+			OPL3_GenerateStream(&chip, buf, todo);
+			chan->AddSamples_s16(todo, buf);
```

Comment:
Suggest moving `samples -= todo` after the two actions that generate and deliver the samples.
(I confirmed that the two downstream functions don't access the `samples` value, and `todo` isn't updated or modified either - so this should be a safe logic improvement)